### PR TITLE
[GPU] Add NV12toRGB/NV12toBGR operations

### PR DESCRIPTION
**Pull request by Lyamin-Roman**
### Tickets:
 - *69187*

### CODE REVIEW

File: inference-engine/tests/functional/plugin/gpu/remote_blob_tests/cldnn_remote_blob_tests.cpp [hunk 1]
```diff
@@ -498,6 +498,142 @@ TEST_P(BatchedBlob_Test, canInputNV12) {
...
+
+        // This network generates [1, size] tensor whether batch=1 or 2. So need to split
+        auto split_shared_blob = make_shared_blob<float_t>(output_blob_local->getTensorDesc(),
+                                    outputBlob_shared->buffer().as<float_t *>() + output_blob_local->size() * i);
+        ASSERT_EQ(output_blob_local->size(), split_shared_blob->size());
+        float thr = 0.1;
+
+        FuncTestUtils::compareBlobs(output_blob_local, split_shared_blob, thr, "", false);
+    }
+}
```
Comment (ilya-lavrenov) [inline inference-engine/tests/functional/plugin/gpu/remote_blob_tests/cldnn_remote_blob_tests.cpp:635]: can we write tests on new OV 2.0 API?

File: inference-engine/src/cldnn_engine/cldnn_common_utils.h [hunk 2]
```diff
@@ -185,4 +185,11 @@ inline std::vector<uint16_t> ConvertPermuteOrder(const std::vector<uint16_t>& ie
     return cldnn_order;
 }
 
+inline InferenceEngine::Layout InferenceEngineLayoutFromOVLayout(ov::Layout l) {//not finished
+    std::string layout = l.to_string();
+    if (layout == "[N,H,W,C]") return InferenceEngine::Layout::NHWC;
```
Comment (nosovmik) [inline inference-engine/src/cldnn_engine/cldnn_common_utils.h:190]: It is better to not rely on internal representation of 'to_string' (it is more for debugging).
Try using `if (layout == Layout("NHWC")` instead

File: inference-engine/thirdparty/clDNN/tests/test_cases/convert_color_gpu_test.cpp [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+
+    auto outputs = network.execute();
+
+    std::vector<float> reference_results(width * height * 3);
+    createReferenceData<uint8_t, float>(input_data.data(), input_data.data() + height * width, reference_results.data(),
+                                      1, height, width, input_height * width, input_height * width, true);
+
+    auto output_prim = outputs.begin()->second.get_memory();
+    cldnn::mem_lock<float> output_ptr(output_prim, get_test_stream());
+    int size = width * height * 3;
```
[the remark this answers is not in the document]
Comment (lznamens) [reply inference-engine/thirdparty/clDNN/tests/test_cases/convert_color_gpu_test.cpp:524]: Or better `reference_results.size()`

File: inference-engine/thirdparty/clDNN/tests/test_cases/convert_color_gpu_test.cpp [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+
+            float B = (1.164f * (float)(y_comp - 16) + 1.596f * (float)(v_comp - 128));
+            float G = (1.164f * (float)(y_comp - 16) - 0.813f * (float)(v_comp - 128) - 0.391f * (u_comp - 128));
+            float R = (1.164f * (float)(y_comp - 16) + 2.018f * (float)(u_comp - 128));
+
+            R = std::min(std::max(R, 0.f), 255.f);
+            G = std::min(std::max(G, 0.f), 255.f);
+            B = std::min(std::max(B, 0.f), 255.f);
+
+            if (format == cldnn::format::bfyx) {
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/tests/test_cases/convert_color_gpu_test.cpp:386]: Is it really needed? I don't see support of bfyx into the implementation_map related to the primitive.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convert_color_ref.cl [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+#ifdef CONVERT_FROM_NV12
+#ifdef BUFFER_MEM
+KERNEL(convert_color_ref)(const __global INPUT0_TYPE* input_y,
+#if INPUTS_COUNT == 2
+                          const __global INPUT1_TYPE* input_uv,
+#endif
+                          __global OUTPUT_TYPE* output) {
+
+    const uint b = get_global_id(0);
+    const uint f = get_global_id(1);
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convert_color_ref.cl:17]: I suppose currently you don't use this dimension. Maybe it's better to use gws = { x, y, b }?
If get_global_size(1) > 1 then you have several parallel threads which make one job.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convert_color/convert_color_kernel_base.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    return true;
+}
+
+CommonDispatchData ConvertColorKernelBase::SetDefault(const convert_color_params& params, const optional_params&) const {
+    const auto& out = params.output;
+    CommonDispatchData dispatchData;
+
+    dispatchData.gws = { out.Batch().v, out.Feature().v, out.Y().v * out.X().v };
+    dispatchData.lws = GetOptimalLocalWorkGroupSizes(dispatchData.gws, params.engineInfo);
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convert_color/convert_color_kernel_base.cpp:30]: It's better from perf point to add info about in and out layout into `GetOptimalLocalWorkGroupSizes`, please find examples in the code.

File: inference-engine/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp [hunk 7]
```diff
@@ -408,11 +408,11 @@ std::vector<std::pair<ngraph::element::Type, std::vector<std::uint8_t>>> LayerTe
...
-        for (const auto &output : outputsInfo) {
-                convertType.push_back(
-                    FuncTestUtils::PrecisionUtils::convertIE2nGraphPrc(
-                        output.second->getTensorDesc().getPrecision()));
-        }
+    for (const auto &output : outputsInfo) {
+            convertType.push_back(
+                FuncTestUtils::PrecisionUtils::convertIE2nGraphPrc(
+                    output.second->getTensorDesc().getPrecision()));
+    }
```
Comment (lznamens) [inline inference-engine/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp:415]:
```suggestion
    for (const auto &output : outputsInfo) {
        convertType.push_back(
            FuncTestUtils::PrecisionUtils::convertIE2nGraphPrc(
                output.second->getTensorDesc().getPrecision()));
    }
```

File: inference-engine/src/cldnn_engine/ops/parameter.cpp [hunk 8]
```diff
@@ -174,107 +177,129 @@ static void CreateParameterOp(Program& p, const std::shared_ptr<ngraph::op::v0::
...
-            auto preprocessPrimID = "reorder:" + inputName + std::to_string(i) + Program::m_preProcessTag;
-            std::string y_name = inputName + "_Y" + std::to_string(i);
-            std::string uv_name = inputName + "_UV" + std::to_string(i);
+            for (auto i = 0; i < inputDims[0]; i++) {
+                auto preprocessPrimID = "reorder:" + inputName + std::to_string(i) + Program::m_preProcessTag;
+                std::string y_name = inputName + "_Y" + std::to_string(i);
+                std::string uv_name = inputName + "_UV" + std::to_string(i);
+
+                cldnn::layout y_layout(DataTypeFromPrecision(ip),
+                                        cldnn::format::nv12, { 1, 1, width, height });
```
Comment (lznamens) [inline inference-engine/src/cldnn_engine/ops/parameter.cpp:219]:
```suggestion
                cldnn::layout y_layout(DataTypeFromPrecision(ip),
                                       cldnn::format::nv12, { 1, 1, width, height });
```

File: inference-engine/src/cldnn_engine/ops/parameter.cpp [hunk 9]
```diff
@@ -174,107 +177,129 @@ static void CreateParameterOp(Program& p, const std::shared_ptr<ngraph::op::v0::
...
-        // for NV12, create two input layouts with reorder instead of one,
-        // and then would expect compound blob in inferRequest
-        if (Layout::NCHW != l &&
-            (Precision::I8 != ip || Precision::U8 != ip)) {
-            IE_THROW() << "Unsupported layout (" << l << ") or precision "
-                               << ip.name() << ") for NV12 input " + inputInfo->name();
+    bool is_convert_color_input = false;
+    for (auto& node : op->get_users()) {
+        std::string node_type = node->get_type_name();
+        is_convert_color_input |= node_type == "NV12toRGB" || node_type == "NV12toBGR";//another way?
```
Comment (lznamens) [inline inference-engine/src/cldnn_engine/ops/parameter.cpp:183]: My point that it's OK.

**Review by lznamens — CHANGES_REQUESTED** [review]

File: inference-engine/thirdparty/clDNN/tests/test_cases/convert_color_gpu_test.cpp [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+    }
+}
+
+struct OpenCL {
+    cl::Context _context;
+    cl::Device _device;
+    cl::CommandQueue _queue;
+
+    OpenCL() {
+        // get Intel iGPU OCL device, create context and queue
```
[the remark this answers is not in the document]
Comment (Lyamin-Roman) [reply inference-engine/thirdparty/clDNN/tests/test_cases/convert_color_gpu_test.cpp]: Looks like this is also valid for dGPU case

File: inference-engine/src/cldnn_engine/ops/convert_color.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+//
+
+#include "cldnn_program.h"
+#include "cldnn_common_utils.h"
+
+#include "cldnn/primitives/convert_color.hpp"
+
+namespace CLDNNPlugin {
+
+static void CreateNV12toRGBOp(Program& p, const std::shared_ptr<ngraph::op::v8::NV12toRGB>& op) {
```
Comment (vladimir-paramuzov) [inline inference-engine/src/cldnn_engine/ops/convert_color.cpp:12]: CreateNV12toRGBOp and CreateNV12toBGROp looks pretty much the same. Could you add common function parameterized with input and output color format?

File: inference-engine/src/cldnn_engine/ops/parameter.cpp (hunk 9 above)
Comment (vladimir-paramuzov) [reply inference-engine/src/cldnn_engine/ops/parameter.cpp:183]: `ngraph::is_type<ngraph::op::v8::NV12toRGB>() || ...`

File: src/plugins/intel_gpu/src/plugin/ops/parameter.cpp [hunk 12]
```diff
@@ -174,107 +179,130 @@ static void CreateParameterOp(Program& p, const std::shared_ptr<ngraph::op::v0::
...
+                            op->output(0).get_rt_info()[ov::preprocess::TensorInfoMemoryType::get_type_info_static()]);
+            if (mem_type->get() == GPUConfigParams::KEY_GPU_SURFACE) {
+                networkInputLayout.format = cldnn::format::nv12;
+            }
         }
-        int height = inputDims[2];
-        int width = inputDims[3];
-        std::vector<cldnn::primitive_id> reorders;
+        networkInputLayout.size = { TensorValue(inputDims[0]), TensorValue(inputDims[3]),
+                                    TensorValue(inputDims[2]), TensorValue(inputDims[1]) };
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/plugin/ops/parameter.cpp:198]: Shouldn't it be like this for `byxf` format?
```
        networkInputLayout.size = { TensorValue(inputDims[0]), TensorValue(inputDims[2]),
                                    TensorValue(inputDims[3]), TensorValue(inputDims[1]) };
```

Comment (Lyamin-Roman) [reply src/plugins/intel_gpu/src/plugin/ops/parameter.cpp:198]: No, its right
inputDims stores data in byxf format
Constructor cldnn::tensor requires bfxy format

**Review by sshlyapn — APPROVED** [review]

File: inference-engine/tests/functional/plugin/gpu/remote_blob_tests/gpu_remote_tensor_tests.cpp [hunk 13]
```diff
@@ -521,119 +507,188 @@ TEST_P(OVRemoteTensorBatched_Test, DISABLED_canInputNV12) {
...
-    }
+    // ------------------------------------------------------
+    // regular inference
+    auto fn_ptr_regular = ngraph::builder::subgraph::makeConvPoolRelu({1, 3, height, width});
 
-    inf_req_remote.infer();
+    using namespace ov::preprocess;
+    auto p_reg = PrePostProcessor(fn_ptr_regular);
+    p_reg.input().tensor().set_element_type(ov::element::u8)
+                      .set_color_format(ov::preprocess::ColorFormat::NV12_TWO_PLANES, {"y", "uv"})
```
Comment (lznamens) [inline inference-engine/tests/functional/plugin/gpu/remote_blob_tests/gpu_remote_tensor_tests.cpp:587]: nit: alignment

**Review by lznamens — APPROVED** [review]
LGTM

File: inference-engine/tests/functional/shared_test_classes/src/single_layer/convert_color_nv12.cpp [hunk 14]
```diff
@@ -44,7 +45,9 @@ void ConvertColorNV12LayerTest::SetUp() {
     } else {
         auto uvShape = ov::Shape{inputShape[0], inputShape[1] / 2, inputShape[2] / 2, 2};
         auto param_y = std::make_shared<ov::op::v0::Parameter>(ngPrc, inputShape);
+        param_y->set_layout(ov::Layout("NHWC"));
         auto param_uv = std::make_shared<ov::op::v0::Parameter>(ngPrc, uvShape);
+        param_uv->set_layout(ov::Layout("NHWC"));
```
Comment (ilya-lavrenov) [inline inference-engine/tests/functional/shared_test_classes/src/single_layer/convert_color_nv12.cpp:50]: CC @nosovmik 
why do we need it?

Comment (Lyamin-Roman) [reply inference-engine/tests/functional/shared_test_classes/src/single_layer/convert_color_nv12.cpp:50]: I haven't found another way to find out input data format

Comment (ilya-lavrenov) [reply inference-engine/tests/functional/shared_test_classes/src/single_layer/convert_color_nv12.cpp:50]: but why do you need this information?

Comment (nosovmik) [reply inference-engine/tests/functional/shared_test_classes/src/single_layer/convert_color_nv12.cpp:50]: According to NV12toRGB/BGR - input shape shall be always in NHWC format.
https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/image/NV12toRGB_8.md
If GPU needs such information, they can rely on this. Anyway input for this operation may not be always a 'Parameter', but some other operation which doesn't have 'layout' attribute

**Review by ilyachur — APPROVED** [review]
nGraph part LGTM

**Pull request merged**